### PR TITLE
Fixed #22449 -- Colorized output of test results

### PR DESCRIPTION
**Pull request by dizballanze**
Hi,

Here is code with tests and documentation for colorizing output of the `manage.py test` command.

Some features:

-  highlighting tracebacks if pygments library is available
-  highlighting SQL queries if pygments library is available and `manage.py test` command was invoked with `--debug-sql` argument
-  highlighting with corresponding colors based on the used pallete (dark or light)
-  can be switched off with `--no-color` argument
-  switching off automatically if the terminal doesn't support colors

Few screenshots:

image

image

### CODE REVIEW

**Comment by Miserlou** [comment]
So excited for this!

Because of the displaying for tagging that just landed in https://github.com/django/django/pull/9664 I think this has a small conflict. It'd be great if we could use this new fancy color output to highlight the included/excluded tags as well!

**Comment by dizballanze** [comment]
So, What should I do now? Should I resolve conflict?

**Comment by Miserlou** [comment]
Yep, just resolve the conflict by pulling in from the main branch and making your changes to it and adding that commit to this branch, which will update the PR.

**Comment by dizballanze** [comment]
Hi @carltongibson.
Thanks for the review.

I think I fixed the issue with the environment variable `DJANGO_COLORS`. Please, try again.

**Comment by carltongibson** [comment]
Hi @dizballanze. 

Some progress. 

With both `DJANGO_COLORS` settings, I'm now just getting the 2 failures in `TestDebugSQL`.

(Previously  `DJANGO_COLORS="light"` was also failing in `TestColoredRunner`) 

You squashed your changes so I can't see what you did... but more of that it looks like 😃

**Comment by dizballanze** [comment]
I think I get it now. Failures occur if `pygments` package is installed.
I will add `skipIf` for this tests.

**Comment by carltongibson** [comment]
Ah, good. So I install Pygments to try the patch, and so break the tests. Nice. 😃

```
pip uninstall pygments
...
django/tests (pr/9688)$ ./runtests.py test_runner
...
----------------------------------------------------------------------
Ran 59 tests in 2.394s

OK (skipped=2)
```

I'll just give it a go with the `skipIf`.

**Comment by dizballanze** [comment]
It should work now.

**Comment by carltongibson** [comment]
Ok. Yep. It skips 4 with pygments installed. 

I'll unflag Patch needs improvement and we'll have another look. 

Good work. Thanks!

File: django/test/runner.py [hunk 1]
```diff
@@ -5,30 +5,226 @@
...
 
 try:
     import tblib.pickling_support
 except ImportError:
     tblib = None
 
 
-class DebugSQLTextTestResult(unittest.TextTestResult):
-    def __init__(self, stream, descriptions, verbosity):
+TERMINATOR = '\x1b[{}m'.format(RESET)
```
Comment (carltongibson) [inline django/test/runner.py]: This is only used once. Can we move it back to the `color` module? (That way `termcolors` is still only ever used by `color`)

File: django/test/runner.py [hunk 2]
```diff
@@ -5,30 +5,226 @@
...
+            self.stream.writeln(" (%s)" % (", ".join(infos),))
+        else:
+            self.stream.write("\n")
+        return result
+
+    def _makeResult(self):
+        return self.resultclass(self.stream, self.descriptions, self.verbosity, no_color=self.no_color)
+
+
+def get_sql_highlighter(bg='light'):
```
Comment (carltongibson) [inline django/test/runner.py]: Can we move this function up by `get_traceback_highlighter`. (That was we keep the two related helper functions next to each other.)

File: django/test/runner.py [hunk 3]
```diff
@@ -5,30 +5,226 @@
...
 except ImportError:
     tblib = None
 
 
-class DebugSQLTextTestResult(unittest.TextTestResult):
-    def __init__(self, stream, descriptions, verbosity):
+TERMINATOR = '\x1b[{}m'.format(RESET)
+
+
+def get_traceback_highlighter(bg='light'):
```
Comment (carltongibson) [inline django/test/runner.py]: This is only ever called once. Do we need the default? 
(Same with SQL version)

File: django/utils/termcolors.py [hunk 4]
```diff
@@ -166,15 +184,17 @@ def parse_color_setting(config_string):
         'bold', 'underscore', 'blink', 'reverse', 'conceal', 'noreset'
     """
     if not config_string:
-        return PALETTES[DEFAULT_PALETTE]
+        return (DEFAULT_PALETTE, PALETTES[DEFAULT_PALETTE])
```
Comment (carltongibson) [inline django/utils/termcolors.py]: I think I'd add a note on the return 2-tuple here to the docstring.

File: django/test/runner.py (hunk 2 above)
Comment (dizballanze) [reply django/test/runner.py]: I think current function location is better because it's closer to where it used.

Comment (carltongibson) [reply django/test/runner.py]: 😃 Yeah, I can see that. 

But I was scanning the code and looking for the definition. I want to find all the helpers in one place. 

(As ever, there are different ways of wanting to read it. That's OK.) 

If you feel strongly you can leave it for now. Tim will have a look before merging it anyway so he can input.

**Comment by dizballanze** [comment]
Hi @carltongibson 

I fixed some of the issues and have a question for [one of them](https://github.com/django/django/pull/9688#discussion_r170595636).

About pygments. It highlights SQL code and I think specific colors can depend on the pygments version. I can add tests to check if test runner output contains highlighted SQL and tracebacks with currently installed pygments library. But we'll need to skip this tests if pygments isn't installed as well.

File: django/test/runner.py (hunk 2 above)
Comment (dizballanze) [reply django/test/runner.py]: No problem, I will move it.

**Comment by carltongibson** [comment]
Hi @dizballanze. 

> About pygments...

Hmmm. OK. I'll go with "leave it" for now. (Someone else can say otherwise.) We don't have `pygments` CI, so the two skipped functions get run there, even if we skip then locally. 

(Looking at it we could add a lot of boilerplate to test not very much here: I think it's safe for us to assume that `pygments` works — so there's no point calling the highlighter just to test the output of the highlighter. We could test that the highlighter function gets called, but that's already exercised, etc...)

The CI didn't seem to run again after https://github.com/django/django/pull/9688/commits/1b16d8bb8a0ae39f7b8df2476ea6a45abd73cbcc. Can you squash these commits (including https://github.com/django/django/pull/9688/commits/a9e336c4861771c61885d8df4650c4e32b5ff049) and push again? (That should restart it.) 

Thanks again. Great stuff!

**Comment by carltongibson** [comment]
(Ah, CI, kicked in.)

**Comment by dizballanze** [comment]
I've squashed commits.

Thanks for the review 👍

**Comment by timgraham** [comment]
I'm not sure that including this in Django is a good idea. In particular, copying entire parts of Python like `TextTestRunner`'s `run()` method could be a maintenance headache. I found some third party solutions like [redgreenunittests](https://github.com/stevematney/redgreenunittests) (which does the same I guess, but since it's a separate package, it can adapt more quickly than Django can).

**Comment by timgraham** [comment]
Closing based on the discussion in the ticket.

**Comment by Miserlou** [comment]
What ticket? What discussion? You haven't provided a link. Or do you mean your own comment where you single-mindedly decided that this feature was out because you said so?

This is a feature that Django should obviously have and it's wrong to suggest that because there is a single, obscure, maintained, many-years-old third party library that we can abandon implementing this.

If you don't like the implementation because there are upstream issues, then you should file those issues before closing an open PR without any further discussion.

**Comment by timgraham** [comment]
Each pull request has a corresponding Trac ticket. The ticket number follows "Fixed #" in the commit message and pull request title. https://code.djangoproject.com/ticket/22449

**Comment by limedaniel** [comment]
Could this code be repackaged into a module like `django-test-colours` or something?  Surely installation would be as easy as `pip install modulename` and then editing `settings.py` to set the test runner.

**Pull request closed without merging**